mitiq/pec/sampling.py

```diff
@@ -121,13 +123,14 @@ def sample_circuit(
             input circuit. If a representation cannot be found for an operation
             in the circuit, a ValueError is raised.
         random_state: Seed for sampling.
-        num_samples: The number of samples.
+        num_samples: The number of samples,
+        extra_data: Mutable dictionary to save extra data beyond output
+            circuits. After the function call it will contain ``signs`` (the
+            signs associated to sampled_circuits) and ``norm`` (the one-norm
+            of the full circuit representation).
```

Comment:
This is an interesting way that I've never considered to give users additional info from a function. I'm accustomed to having a return type of something like `Union[float, Tuple[float, Dict[str, int]]]` where `float` is returned by default, and `Tuple[float, Dict[str, int]]` is returned if some flag is passed as `True` (or similar).

Interesting idea!

What — in your mind — are the tradeoffs of using this method, over the more "traditional" method of returning extra data if the user asks?

Reply:
I have seen this approach in other contexts. The first time I have seen this was on C++ OpenCV functions :smile:    see e.g. the output of the [blur](https://docs.opencv.org/3.4/d4/d86/group__imgproc__filter.html#ga8c45db9afe636703801b0b2e440fce37)() function.

Personally, I find it a bit counter-intuitive but there are cases in which it can be useful (my debatable opinions):
 
- For logging information on a mutable "recording" variable passing through multiple functions. I think it's handy to just pass to the function the previous logged data and ask, "please continue appending info to this object".
- If the mutable argument is optional, it simplifies the structure of the returned type. Avoiding complicated things like: 
```
if return_all: 
    return y, data 
else 
    return y
```
- In cases in which the function return type cannot be changed, in order to interfacing well with other functions or decorators.  This is exactly the case of this PR. The conversion decorator expects a Cirq function returning `List[cirq.Circuit]` but we also wanted to output additional data: `signs` and `norm`. So using the `extra_data` argument in this case was quite helpful for making the function compatible with the decorator.


Having said that, I don't know it this is a good practice.

Reply:
Thanks Andrea for the thoughts. I agree that in certain scenarios this pattern seems acceptable (maybe even good!). If for some reason down the line we run into problems, we can always modify it.